zitadel/v2/org/funcs.go

```diff
@@ -82,47 +74,92 @@ func update(ctx context.Context, d *schema.ResourceData, m interface{}) diag.Dia
-	tflog.Debug(ctx, "check if org is existing", map[string]interface{}{
-		"id": orgID,
+	resp, err := client.GetOrgByID(ctx, &admin.GetOrgByIDRequest{
+		Id: orgID,
 	})
-
-	for i := range resp.Result {
-		org := resp.Result[i]
-		if strings.Compare(org.GetId(), orgID) == 0 {
-			d.SetId(orgID)
-			tflog.Debug(ctx, "found org", map[string]interface{}{
-				"id": orgID,
-			})
-			if err := d.Set(nameVar, org.GetName()); err != nil {
-				return diag.Errorf("failed to set %s of org: %v", nameVar, err)
-			}
-			return nil
-		}
+	if err != nil {
+		return diag.Errorf("error while getting org by id %s: %v", orgID, err)
+	}
+	remoteOrg := resp.GetOrg()
+	d.SetId(remoteOrg.Id)
+	if err := d.Set(nameVar, remoteOrg.Name); err != nil {
+		return diag.Errorf("error while setting org name %s: %v", remoteOrg.Name, err)
+	}
+	if err := d.Set(primaryDomainVar, remoteOrg.PrimaryDomain); err != nil {
+		return diag.Errorf("error while setting org primary domain %s: %v", remoteOrg.PrimaryDomain, err)
+	}
+	state := org.OrgState_name[int32(remoteOrg.State)]
+	if err := d.Set(stateVar, state); err != nil {
+		return diag.Errorf("error while setting org state %s: %v", state, err)
 	}
-
-	d.SetId("")
-	tflog.Debug(ctx, "org not found", map[string]interface{}{})
 	return nil
 }
+
+func list(ctx context.Context, d *schema.ResourceData, m interface{}) diag.Diagnostics {
+	tflog.Info(ctx, "started list")
+	orgName := d.Get(nameVar).(string)
+	orgNameMethod := d.Get(nameMethodVar).(string)
+	orgDomain := d.Get(domainVar).(string)
+	orgDomainMethod := d.Get(domainMethodVar).(string)
+	orgState := d.Get(stateVar).(string)
+	clientinfo, ok := m.(*helper.ClientInfo)
+	if !ok {
+		return diag.Errorf("failed to get client")
+	}
+	client, err := helper.GetAdminClient(clientinfo)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+	req := &admin.ListOrgsRequest{}
+	if orgName != "" {
+		req.Queries = append(req.Queries, &org.OrgQuery{
+			Query: &org.OrgQuery_NameQuery{
+				NameQuery: &org.OrgNameQuery{
+					Name:   orgName,
+					Method: object.TextQueryMethod(object.TextQueryMethod_value[orgNameMethod]),
+				},
+			},
+		})
+	}
+	if orgState != "" {
+		req.Queries = append(req.Queries, &org.OrgQuery{
+			Query: &org.OrgQuery_StateQuery{
+				StateQuery: &org.OrgStateQuery{
+					State: org.OrgState(org.OrgState_value[orgState]),
+				},
+			},
+		})
+	}
+	if orgDomain != "" {
+		req.Queries = append(req.Queries, &org.OrgQuery{
+			Query: &org.OrgQuery_DomainQuery{
+				DomainQuery: &org.OrgDomainQuery{
+					Domain: orgDomain,
+					Method: object.TextQueryMethod(object.TextQueryMethod_value[orgDomainMethod]),
+				},
+			},
+		})
+	}
+	resp, err := client.ListOrgs(ctx, req)
+	if err != nil {
+		return diag.Errorf("error while getting org by id %s: %v", orgName, err)
+	}
+	orgIDs := make([]string, len(resp.Result))
+	for i, org := range resp.Result {
+		orgIDs[i] = org.Id
+	}
+	// If the ID is blank, the datasource is deleted and not usable.
+	d.SetId("-")
```

Comment:
Did you test that if the list of orgs changes that the logic with for each gets managed correctly with terraform?

Reply:
I tested this config:

```hcl
data zitadel_orgs list {
  domain = "policy"
  domain_method = "TEXT_QUERY_METHOD_CONTAINS_IGNORE_CASE"
}

data zitadel_org get {
  for_each = toset(data.zitadel_orgs.list.ids)
  id = each.value
}

output "name" {
  value = toset([
    for org in data.zitadel_org.get : org.name
  ])
}
```

When adding an org matching the criteria, the new org is added to the plan:

```
tf apply
data.zitadel_orgs.list: Reading...
data.zitadel_orgs.list: Read complete after 0s [id=-]
data.zitadel_org.get["227660920908677123"]: Reading...
data.zitadel_org.get["227657510687866883"]: Reading...
data.zitadel_org.get["227777495313678339"]: Reading...
data.zitadel_org.get["227657510687866883"]: Read complete after 0s [id=227657510687866883]
data.zitadel_org.get["227660920908677123"]: Read complete after 0s [id=227660920908677123]
data.zitadel_org.get["227777495313678339"]: Read complete after 0s [id=227777495313678339]

Changes to Outputs:
  ~ name = [
        # (1 unchanged element hidden)
        "domain-policy-org-jfftgbbewx",
      + "new-policy",
    ]

You can apply this plan to save these new output values to the Terraform state, without changing any real infrastructure.

Do you want to perform these actions?
  Terraform will perform the actions described above.
  Only 'yes' will be accepted to approve.

  Enter a value: yes


Apply complete! Resources: 0 added, 0 changed, 0 destroyed.

Outputs:

name = toset([
  "domain-policy-org-dxmnvkusoo",
  "domain-policy-org-jfftgbbewx",
  "new-policy",
])
```